r/aws_appautoscaling_target: Support updating max_capacity, min_capacity, and role_arn attributes

bflad opened the pull request:
The [RegisterScalableTarget](http://docs.aws.amazon.com/sdk-for-go/api/service/applicationautoscaling/#ApplicationAutoScaling.RegisterScalableTarget) API call supports updating attributes on existing targets.

Closes #240 and potentially #968 

```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSAppautoScalingTarget'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSAppautoScalingTarget -timeout 120m
=== RUN   TestAccAWSAppautoScalingTarget_basic
--- PASS: TestAccAWSAppautoScalingTarget_basic (67.34s)
=== RUN   TestAccAWSAppautoScalingTarget_spotFleetRequest
--- PASS: TestAccAWSAppautoScalingTarget_spotFleetRequest (69.89s)
=== RUN   TestAccAWSAppautoScalingTarget_emrCluster
--- PASS: TestAccAWSAppautoScalingTarget_emrCluster (556.08s)
=== RUN   TestAccAWSAppautoScalingTarget_multipleTargets
--- PASS: TestAccAWSAppautoScalingTarget_multipleTargets (32.41s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws   725.772s
```

Ninir commented on aws/resource_aws_appautoscaling_target.go:
```diff
@@ -118,6 +116,36 @@ func resourceAwsAppautoscalingTargetRead(d *schema.ResourceData, meta interface{
+	}
+
+	if d.HasChange("role_arn") {
+		input.RoleARN = aws.String(d.Get("role_arn").(string))
```
Can you add the role_arn update in tests so that we ensure it's all ok?
Seems rather good for me otherwise :)

bflad commented:
@Ninir updated! Thanks for noticing that. I changed the test to actually update the IAM role and it triggered an IAM eventual consistency issue, which is now fixed with retries:
```
        * aws_appautoscaling_target.bar: 1 error(s) occurred:

        * aws_appautoscaling_target.bar: ValidationException: Unable to assume IAM role: arn:aws:iam::123456789012:role/autoscaleroleupdatecluster-tyavrhv2bh
            status code: 400, request id: eef62b85-deb0-11e7-84a9-6b6b83f6f609
```

```
 make testacc TEST=./aws TESTARGS='-run=TestAccAWSAppautoScalingTarget'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSAppautoScalingTarget -timeout 120m
=== RUN   TestAccAWSAppautoScalingTarget_basic
--- PASS: TestAccAWSAppautoScalingTarget_basic (91.72s)
=== RUN   TestAccAWSAppautoScalingTarget_spotFleetRequest
--- PASS: TestAccAWSAppautoScalingTarget_spotFleetRequest (59.53s)
=== RUN   TestAccAWSAppautoScalingTarget_emrCluster
--- PASS: TestAccAWSAppautoScalingTarget_emrCluster (2045.85s)
=== RUN   TestAccAWSAppautoScalingTarget_multipleTargets
--- PASS: TestAccAWSAppautoScalingTarget_multipleTargets (36.86s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	2234.006s
```

radeksimko commented on aws/resource_aws_appautoscaling_target.go:
```diff
@@ -74,8 +72,8 @@ func resourceAwsAppautoscalingTargetCreate(d *schema.ResourceData, meta interfac
 		if err != nil {
-			if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "ValidationException" {
-				log.Printf("[DEBUG] Retrying creation of Application Autoscaling Scalable Target due to possible issues with IAM: %s", awsErr)
+			if isAWSErr(err, "ValidationException", "Unable to assume IAM role") {
```
👍 More precise matching, yey

bflad commented:
Given the large amount of headache with the newly forced IAM service roles, I'm going to resubmit this PR, likely without `role_arn` bits.
